### PR TITLE
Add telemetry for auto-add package owner feature

### PR DESCRIPTION
**Pull request by xavierdecoster**

### CODE REVIEW

File: src/NuGetGallery/Services/TelemetryService.cs [hunk 1]
```diff
@@ -310,6 +312,16 @@ public void TrackPackageRevalidate(Package package)
...
         }
 
+        public void TrackPackageFailedMetadataCompliance(string packageId, string packageVersion)
+        {
+            TrackMetricForPackage(Events.PackageFailedMetadataCompliance, packageId, packageVersion);
+        }
+
+        public void TrackPackageOwnershipAutomaticallyAdded(string packageId, string packageVersion, int userKey)
+        {
+            TrackMetricForPackage(Events.PackageOwnershipAutomaticallyAdded, packageId, packageVersion, properties => properties.Add("RequiredCoOwnerUserKey", userKey.ToString()));
```
Comment (loic-sharma) [inline src/NuGetGallery/Services/TelemetryService.cs]: Let's make `RequiredCoOwnerUserKey` a public const field to stay consistent with all of the other metrics.

Comment (xavierdecoster) [reply src/NuGetGallery/Services/TelemetryService.cs]: Missed that. Fixing.

File: src/NuGetGallery/Services/TelemetryService.cs [hunk 2]
```diff
@@ -310,6 +318,24 @@ public void TrackPackageRevalidate(Package package)
             TrackMetricForPackage(Events.PackageRevalidate, package);
         }
 
+        public void TrackPackageFailedMetadataCompliance(string packageId, string packageVersion, IEnumerable<string> complianceFailures)
+        {
+            TrackMetricForPackage(
+                Events.PackageFailedMetadataCompliance, 
+                packageId, 
+                packageVersion, 
+                properties => properties.Add(ComplianceFailures, string.Join(Environment.NewLine, complianceFailures)));
```
Comment (joelverhagen) [inline src/NuGetGallery/Services/TelemetryService.cs]: I think you can serialize a collection a JSON. AI should be able to parse it.

Comment (xavierdecoster) [reply src/NuGetGallery/Services/TelemetryService.cs]: Cool! Updated.

**Comment by karann-msft** [comment]
How do I access the data from this telemetry?

File: src/NuGetGallery/Security/RequirePackageMetadataCompliancePolicy.cs [hunk 3]
```diff
@@ -60,6 +63,9 @@ public override async Task<SecurityPolicyResult> EvaluateAsync(PackageSecurityPo
             // Automatically add the required co-owner when metadata is compliant.
             if (!context.Package.PackageRegistration.Owners.Select(o => o.Username).Contains(state.RequiredCoOwnerUsername, StringComparer.OrdinalIgnoreCase))
             {
+                // Telemetry
+                context.TelemetryService.TrackPackageOwnershipAutomaticallyAdded(context.Package.Id, context.Package.NormalizedVersion, requiredCoOwner.Key);
```
Comment (skofman1) [inline src/NuGetGallery/Security/RequirePackageMetadataCompliancePolicy.cs]: Feedback from Karan: we shouldn't log Microsoft user key, but log the user key for the team pushing the package (example: if Karan is a collaborater in organization 'Asp.Net', please log 'Asp.Net' key).

**Comment by skofman1** [comment]
>                 return SecurityPolicyResult.CreateWarningResult(Strings.SecurityPolicy_RequirePackagePrefixReserved);

Feedback from Karan: please also add telemetry for cases when we send a warning e-mail. The data should contain package id, version, user key

---
Refers to: src/NuGetGallery/Security/RequirePackageMetadataCompliancePolicy.cs:78 in 86b192b. [](commit_id = 86b192b151634d748a8888d624ba13fcc0f1a450, deletion_comment = False)

File: src/NuGetGallery/Security/RequirePackageMetadataCompliancePolicy.cs (hunk 3 above)
Comment (xavierdecoster) [reply src/NuGetGallery/Security/RequirePackageMetadataCompliancePolicy.cs]: Replaced by adding source and target account to telemetry event.

**Comment by xavierdecoster** [comment]
Now also emitting PackageMetadataComplianceWarning event.

**Comment by xavierdecoster** [comment]
@karann-msft In Application Insights, you'll be able to query for `PackageMetadataComplianceError`, `PackageMetadataComplianceWarning`, and `PackageOwnershipAutomaticallyAdded` events.

File: src/NuGetGallery/Security/RequirePackageMetadataCompliancePolicy.cs [hunk 4]
```diff
@@ -82,6 +84,13 @@ public override async Task<SecurityPolicyResult> EvaluateAsync(PackageSecurityPo
             // the account pushing the package has not registered the prefix yet.
             if (!context.Package.PackageRegistration.IsVerified)
             {
+                // Telemetry
+                context.TelemetryService.TrackPackageMetadataComplianceWarning(
```
Comment (chenriksson) [inline src/NuGetGallery/Security/RequirePackageMetadataCompliancePolicy.cs]: nit: Is there a better name for `TrackPackageMetadataComplianceWarning`? It's not clear from the name that package metadata is valid, but namespace isn't reserved.

Or, if we're adding more warnings... could have `IEnumerable<string> complianceWarnings` argument like we do for failures?

File: src/NuGetGallery/Services/TelemetryService.cs [hunk 5]
```diff
@@ -53,6 +53,7 @@ internal class Events
             public const string SymbolPackagePushFailure = "SymbolPackagePushFailure";
             public const string SymbolPackageGalleryValidation = "SymbolPackageGalleryValidation";
             public const string PackageFailedMetadataCompliance = "PackageFailedMetadataCompliance";
+            public const string PackageMetadataComplianceWarning = "PackageMetadataComplianceWarning";
```
Comment (chenriksson) [inline src/NuGetGallery/Services/TelemetryService.cs]: nit: could have more consistent naming. e.g., `PackageMetadataComplianceFailures`, `PackageMetadataComplianceWarnings`?

File: src/NuGetGallery/Services/TelemetryService.cs [hunk 6]
```diff
@@ -329,13 +330,29 @@ public void TrackPackageRevalidate(Package package)
             TrackMetricForPackage(Events.PackageRevalidate, package);
         }
 
-        public void TrackPackageFailedMetadataCompliance(string packageId, string packageVersion, IEnumerable<string> complianceFailures)
+        public void TrackPackageFailedMetadataCompliance(string packageId, string packageVersion, int sourceAccountKey, int targetAccountKey, IEnumerable<string> complianceFailures)
```
Comment (chenriksson) [inline src/NuGetGallery/Services/TelemetryService.cs]: Aren't acccount keys considered PII. Do we need to exclude them from telemetry? /cc @cristinamanum

File: src/NuGetGallery/Security/RequirePackageMetadataCompliancePolicy.cs (hunk 4 above)
Comment (chenriksson) [reply src/NuGetGallery/Security/RequirePackageMetadataCompliancePolicy.cs]: Looks like I reviewed an  older version of this PR, but I think comments are still relevant. :-)

Comment (xavierdecoster) [reply src/NuGetGallery/Security/RequirePackageMetadataCompliancePolicy.cs]: Added a `IEnumerable<string> complianceWarnings` argument.

File: src/NuGetGallery/Services/TelemetryService.cs (hunk 5 above)
Comment (xavierdecoster) [reply src/NuGetGallery/Services/TelemetryService.cs]: Yep, already handled. `PackageMetadataComplianceError` and `PackageMetadataComplianceWarning` it is.

File: src/NuGetGallery/Services/TelemetryService.cs (hunk 6 above)
Comment (xavierdecoster) [reply src/NuGetGallery/Services/TelemetryService.cs]: Removed PII in this commit https://github.com/NuGet/NuGetGallery/pull/6374/commits/66e4a63be4358d2af5b0b97c4727259a4e595c88

File: src/NuGetGallery/Security/MicrosoftTeamSubscription.cs [hunk 7]
```diff
@@ -45,6 +45,7 @@ private static List<UserSecurityPolicy> InitializePoliciesList()
                     allowedCopyrightNotices: new string[] 
                     {
                         "(c) Microsoft Corporation. All rights reserved.",
+                        "&#169; Microsoft Corporation. All rights reserved.",
```
Comment (chenriksson) [inline src/NuGetGallery/Security/MicrosoftTeamSubscription.cs]: This list seems like a good option for config or resx, stored in a static cache.

File: src/NuGetGallery/Security/RequirePackageMetadataCompliancePolicy.cs [hunk 8]
```diff
@@ -49,6 +49,12 @@ public override async Task<SecurityPolicyResult> EvaluateAsync(PackageSecurityPo
             // Evaluate package metadata validations
             if (!IsPackageMetadataCompliant(context.Package, state, out var complianceFailures))
             {
+                // Telemetry
```
Comment (chenriksson) [inline src/NuGetGallery/Security/RequirePackageMetadataCompliancePolicy.cs]: nit: comment isn't really useful

File: src/NuGetGallery/Security/RequirePackageMetadataCompliancePolicy.cs [hunk 9]
```diff
@@ -60,6 +66,11 @@ public override async Task<SecurityPolicyResult> EvaluateAsync(PackageSecurityPo
...
             if (!context.Package.PackageRegistration.Owners.Select(o => o.Username).Contains(state.RequiredCoOwnerUsername, StringComparer.OrdinalIgnoreCase))
             {
+                // Telemetry
+                context.TelemetryService.TrackPackageOwnershipAutomaticallyAdded(
+                    context.Package.Id, 
+                    context.Package.NormalizedVersion);
+
                 // This will also mark the package as verified if the prefix has been reserved by the co-owner.
                 // The entities context is committed later as a single atomic transaction (see PackageUploadService).
                 await context.PackageOwnershipManagementService.AddPackageOwnerAsync(context.Package.PackageRegistration, requiredCoOwner, commitChanges: false);
```
Comment (cristinamanum) [inline src/NuGetGallery/Security/RequirePackageMetadataCompliancePolicy.cs]: If this line throws. I think that the above telemetry will not be correct. Should the telemetry be added after this line?

Comment (xavierdecoster) [reply src/NuGetGallery/Security/RequirePackageMetadataCompliancePolicy.cs]: Great catch! Fixing

File: src/NuGetGallery/Security/RequirePackageMetadataCompliancePolicy.cs (hunk 8 above)
Comment (xavierdecoster) [reply src/NuGetGallery/Security/RequirePackageMetadataCompliancePolicy.cs]: Fixing

**Pull request merged**